SantaFinder.ts

```diff
@@ -1,110 +1,110 @@
+		while (currentSanta) {		
+			this.availableSantas = this.availableSantas.filter((santa) => { return santa != currentSanta.santa })
+			this.availablePresentees = this.availablePresentees.filter((santa) => { return santa != currentSanta.presentee })
+            this.setChosenPresentee(currentSanta.santa, currentSanta.presentee)
+
+            currentSanta = this.getNextSanta()
+		}
+		
+        // Validity check
+        if (this.availableSantas.length || this.availablePresentees.length) {
+            throw 'UNDEFINED-CIRCLE'
+        }
     }
 
-    chooseRandomCycle() {
-        // Choose random cycle
-        const randomIndex = this.getRandomInt(0, this.possibleCycles.length)
-        const chosenCycle = this.possibleCycles[randomIndex]
-        if (!chosenCycle) throw 'UNDEFINED-CIRCLE'
-
-        chosenCycle.forEach((id, i) => {
-            const presenteeId = (i + 1) === chosenCycle.length ? chosenCycle[0] : chosenCycle[i + 1]
-            this.setChosenPresentee(id, presenteeId)
+    private getNextSanta() {
+        // Find all santas that still need an presentee
+        let availableNodes = this.santas
+            .map((santa) => {
+                if (this.availableSantas.includes(santa.id)) {
+                    const presentees = santa.possiblePresentees.filter((presentee: string) => { return this.availablePresentees.includes(presentee) })
+                    return { id: santa.id, possiblePresentees: presentees }
+                }
+            })
+            .filter((item): item is SantaNode => !!item)
+        
+        // We are done
+        if (availableNodes.length === 0) {
+            return null
+        }
+        
+        // Find presentees who should be prioritized because few santas apply
+        const presenteeCounts = availableNodes.flatMap((node) => { return node.possiblePresentees })
+            .reduce((acc, curr) => {
+                return acc[curr] ? ++acc[curr] : acc[curr] = 1, acc
+            }, {} as {[id: string]: number; })
+        const presentees = Object.entries(presenteeCounts)
+            .sort((p1, p2) => { return p1[1] - p2[1] })
+    
+        // Find santas with least amount of possible presentees
+        availableNodes.sort((node1, node2) => {
+            return node1.possiblePresentees.length - node2.possiblePresentees.length
+        })        
+        let count = availableNodes[0].possiblePresentees.length
+        availableNodes = availableNodes.filter((node) => {
+            return node.possiblePresentees.length === count
         })
-        console.log(chosenCycle)
-    }
 
-    getRandomSantas() {
-        this.chooseRandomCycle()
+        // Choose random santa and presentee (within priority constraints)
+        const random = Math.floor(Math.random() * availableNodes.length);
+        const node = availableNodes[random]
+        const presentee = presentees.filter((presentee) => { return node.possiblePresentees.includes(presentee[0]) })[0]
```

Comment:
```suggestion
        const presentee = presentees.find((presentee) => node.possiblePresentees.includes(presentee))
```
Statt filter()[0] kannst du auch .find() nutzen